hta/analyzers/cupti_counter_analysis.py

```diff
@@ -72,60 +56,41 @@ def _get_counter_data_with_operators_for_rank(
             )
             return None
 
-        # Now, we do the merge using index only
-        gpu_kernels = gpu_kernels.merge(
-            kernel_launches[["index"]],
-            left_index=True,
-            right_index=True,
-            how="inner",
-            suffixes=[None, "_runtime"],  # adds a column index_runtime
-        )
-
-        # Add the op_stack as an array
-        def get_opstack(row: pd.Series) -> List[int]:
-            for i in call_stacks_idxs:
-                op_stack = cg.call_stacks[i].get_path_to_root(row.index_runtime)[:-2]
-                if len(op_stack) > 0:
-                    return op_stack
-            return []
-
-        gpu_kernels["op_stack"] = gpu_kernels.apply(get_opstack, axis=1)
+        gpu_kernels["index_runtime"] = kernel_launches["index"]
+        # Add stack columns
+        op_stacks = {}
+        top_level_ops = {}
+        bottom_level_ops = {}
+        for idx, kernel_idx, launch_idx in gpu_kernels[
+            ["index", "index_runtime"]
+        ].itertuples(index=True):
+            stack = cg.get_stack_of_node(launch_idx).sort_values("ts")
+            ops = stack.loc[stack.s_cat.eq("cpu_op"), "index"].to_list()
```

Comment:
@fengxizhou  I thought stacks contain runtime events too, that is why on the left I skipped one level, that might be wrong assumption, please let me kno

Reply:
Yes. It includes the runtime. However, here we only extract the `cpu_op` events.